### ISSUE TITLE
Id selector parsed in SCSS class selector using interpolation

### ISSUE THREAD

**Issue by alistairjcbrown** [stylelint/stylelint#945]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

When using SCSS string interpolation within a class selector, an ID selector is also parsed due to the `#` in the SCSS string interpolation syntax This is most apparent when validating selectors with regex.

> Which rule, if any, is this issue related to?

Related to `selector-class-pattern` and `selector-id-pattern`, but more related to how selectors are parsed / SCSS is dealt with.

> What CSS is needed to reproduce this issue?

```css
$name: foo;
.#{$name} {
  border-color: blue;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-class-pattern": "^[a-z]+[a-z-_]+[a-z]+$",
    "selector-id-pattern": "^[a-z]+[a-z-_]+[a-z]+$"
  }
}
```

> Which version of stylelint are you using?

Version `4.3.3`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Using `stylelint` with `gulp-postcss`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to SCSS interpolation in selectors

> What did you expect to happen?

Validation on the completed selector (after interpolation, eg. `.foo`), or a warning that SCSS interpolation within selectors is not supported.

> What actually happened (e.g. what warnings or errors you are getting)?

Selector is parsed as CSS, with the `.` section being processed as a class and the `#{$name}` section being processed as an ID. Both of these fail the regex set above, and two errors are output:

```shell
Expected class selector "." to match specified pattern (selector-class-pattern) [stylelint]
Expected id selector "#{$name}" to match specified pattern (selector-id-pattern) [stylelint]
```

**Comment by davidtheclark** [comment]
> Validation on the completed selector (after interpolation, eg. .foo)

The only way you can get this is by compiling your SCSS before running it through stylelint. Neither stylelint nor PostCSS *compile* SCSS --- let's leave that to Sass itself.

(That said, I wonder if they made a poor choice of an interpolation syntax that can be used in selectors, because of problems like this that it causes.)

So I'm not sure you're going to be able to use these rules if you want to use Sass interpolation in your selectors. That said, we'll leave this open for ideas for a little while, if you or anybody else comes up with a solution.

**Comment by TxHawks** [comment]
One way to go about this would be to check if the `#` is directly followed be  `{`, which would mean it's an interpolated value, not an ID

**Comment by alistairjcbrown** [comment]
@davidtheclark That makes sense that it would need to be processed first - as @TxHawks mentions above, would it be possible to detect when string interpolation is occurring in the selector and ignore / output a warning that it can't be validated?

**Comment by davidtheclark** [comment]
The issue is in postcss-selector-parser, so I'll bring it up there.

**Comment by jeddy3** [comment]
Waiting on a release of `postcss-selector-parser`.

**Comment by davidtheclark** [comment]
Just to be clear: this is awaiting the release of postcss-selector-parser 2.0.

**Comment by kboucher** [comment]
I am receiving an `Unknown word CssSyntaxError` when trying to parse files using SASS string interpolation. I am not using the `selector-no-id` rule.

Should I expect this issue to be fixed with `postcss-selector-parser 2.0.`?

Example code

```
$type: 'something';
$size: 10;

.my-selector {
    #{$type}: #{$size};
}
```

Result:
` 5:7  ✖  Unknown word   CssSyntaxError`

**Comment by jeddy3** [comment]
@kboucher This is likely unrelated to `postcss-selector-parser`. Have you [set the syntax option to scss](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md#parsing-non-standard-syntax)?

**Comment by kboucher** [comment]
@jeddy3 Thanks, that did it.

**Comment by isellsoap** [comment]
I’m using stylint 6.9.0 and still have this issue. Can someone confirm?

**Comment by evilebottnawi** [comment]
@isellsoap yes, confirm, i reopen and tomorrow I will try to understand why

**Comment by evilebottnawi** [comment]
@isellsoap can reproduce on new version `stylelint`, try use latest version, maybe also try remove `node_modules` and run again `npm install` and tests.

**Comment by Arcanemagus** [comment]
@evilebottnawi *can't?

**Comment by ntwb** [comment]
> *can't?

I believe so, I edited the comment to change from `can` to `can't` 😄
